sycl/doc/design/DeviceConfigFile.md

```diff
@@ -0,0 +1,297 @@
+  AspectExt_intel_memory_clock_rate = 38,
+  AspectExt_intel_pci_address = 18,
+  AspectExt_oneapi_bfloat16_math_functions = 35,
+  AspectExt_oneapi_cuda_async_barrier = 34,
+  AspectExt_oneapi_native_assert = 31,
+  AspectExt_oneapi_srgb = 30,
+  AspectFp16 = 5,
+  AspectFp64 = 6,
+  AspectGpu = 2,
+  AspectHost_debuggable = 32,
+  AspectImage = 9,
+  AspectOnline_compiler = 10,
+  AspectOnline_linker = 11,
+  AspectQueue_profiling = 12,
+  AspectUsm_atomic_host_allocations = 26,
+  AspectUsm_atomic_shared_allocations = 27,
+  AspectUsm_device_allocations = 13,
+  AspectUsm_host_allocations = 14,
+  AspectUsm_shared_allocations = 15,
+  AspectUsm_system_allocations = 17,
+};
+    
+struct TargetInfo {
+  bool maySupportOtherAspects;
+  std::vector<unsigned> aspects;
+  std::vector<unsigned> subGroupSizes;
+  std::string aotToolchain;
+  std::string oclocDevice;
+};
+
+std::map<std::string, TargetInfo> targets = {
+    {"TargetA",
+     {{AspectCpu, AspectAtomic64}, 0, {8, 16}, "ocloc", "tgtA"}},
+    {"TargetB",
+     {{AspectGpu, AspectFp16}, 0, {8, 16}, "ocloc", "tgtB"}},
+    {"TargetC",
+     {{AspectEmulated, AspectImage}, 0, {8, 32}, "ocloc", "tgtC"}}};
+```
+
+We also need a header file that includes the `.inc` file generated by the
+TableGen backend. Other backends don't generate the definition of `struct
+TargetInfo`. This might be a limitation of the backends, or a limitation of
+TableGen. In case it cannot be autogenerated by the backend, a possible
+workaround is simply to define the struct in this header file. This header
+file should look like the code below:
+```c++
+namespace DeviceConfigFile {
+// In case the backend cannot generate a definition for TargetInfo
+struct TargetInfo {
+  bool maySupportOtherAspects;
+  std::vector<unsigned> aspects;
+  std::vector<unsigned> subGroupSizes;
+  std::string aotToolchain;
+  std::string oclocDevice;
+};
+
+#include "device_config_file.inc"
+}; // namespace DeviceConfigFile
+```
+
+Other modules can query the map to get the information like in the example
+below:
+```c++ 
+DeviceConfigFile::TargetInfo info = DeviceConfigFile::targets.find("TargetA");
+if (info == DeviceConfigFile::targets.end()) {
+  /* Target not found */
+  ...
+} else {
+  auto aspects = info.aspects;
+  auto maySupportOtherAspects = info.maySupportOtherAspects;
+  auto subGroupSizes = info.subGroupSizes;
+  ...
+}
+```
+
+### Changes to Build Infrastructure
+We need the information about the targets in multiple tools and compiler
+modules listed in [Requirements](#Requirements).  Thus, we need to make sure
+that the generation of the `.inc` file out of the `.td` file is done in time
+for all the consumers. The command we need to run for TableGen is `llvm-tblgen
+-gen-dynamic-tables -I /llvm-root/llvm/include/ input.td -o output.inc`.
+Additionally, we need to set dependencies adequately so that this command is
+run before any of the consumers need it.
+
+### Changes to the DPC++ Frontend
```

Comment:
I think it would be good to list all the processes that need to read the config file.  I think it is:

* Compiler driver: in order to know the legal values for `-fsycl-targets`.
* Clang: in order to emit diagnostics related to `-fsycl-fixed-targets`.
* sycl-post-link: in order to filter kernels in device images when doing AOT.